src/FSharp.Data.GraphQL.Server/Executor.fs

```diff
@@ -1,16 +1,22 @@
 ﻿namespace FSharp.Data.GraphQL
 
-open System
 open System.Collections.Generic
 open FSharp.Data.GraphQL.Types
 open FSharp.Data.GraphQL.Execution
 open FSharp.Data.GraphQL.Ast
 open FSharp.Data.GraphQL.Parser
-open FSharp.Data.GraphQL.Types.Patterns
 open FSharp.Data.GraphQL.Planning
 
+type ExecutionFunc<'Root> = 
+    ExecutionPlan * 'Root option * Map<string, obj> option -> Async<GQLResponse>
 
-type Executor<'Root> (schema: ISchema<'Root>) = 
+type ExecutionMiddlewareFunc<'Root> = 
+    ExecutionPlan * 'Root option * Map<string, obj> option -> ExecutionFunc<'Root> -> Async<GQLResponse>
+
+type IExecutionMiddleware<'Root> =
+    abstract member AsyncResolve : ExecutionMiddlewareFunc<'Root>
+
+type Executor<'Root>(schema: ISchema<'Root>, middlewares : IExecutionMiddleware<'Root> seq) = 
```

Comment:
One of the ways to inject middleware is by implementing IExecutionMiddleware<'Root> interfaces for each middleware needed.

The interface does have one single function, AsyncResolve, which receives the arguments of a execution process (plan, root data and variables) and the next execution call (if the middleware needs to pass on to the next middleware).